From 92c99e22979592b2f8808673b540bb7bbb19ed64 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Tue, 07 Dec 2021 08:59:53 +0200
Subject: [PATCH] [SW-63201] gaudi2: security: add psoc rr for lbw

Gaudi2 does not have dedicated lbw range registers for
psoc(jt,i2c,smi) mstr ifs. Instead the hbw range registers
in these mstr ifs can be used for both hbw and lbw purpose.

This is a special case usage only valid for psoc rrs in gaudi2.

Change-Id: Ib03cc50443be3c91fb578ed76eeefb0ecb092561
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index 64db12c..a360e80 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -160,6 +160,13 @@
 	HL_SEC_HBW_ICCM0_TO_RAZWI = 2,
 	HL_SEC_HBW_SRAM_BGN = 3,
 	HL_SEC_HBW_SRAM_IMG = 4,
+	/*
+	 * HL_SEC_HBW_PSOC_LBW is a special case RR, because
+	 * for gaudi2 JT/MSTR/I2C RRs do not have a dedicated
+	 * LBW RR, but instead can use the HBW RRs to block
+	 * LBW Ranges.
+	 */
+	HL_SEC_HBW_PSOC_LBW = 5,
 	HL_SEC_HBW_SRAM = 6,
 	/*
 	 * SRAM END points to the first byte of HBM
@@ -218,15 +225,6 @@
 		.max_full_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_0,
 		.hit_wr_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AW,
 		.hit_rd_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
-	},
-	[HL_SEC_ENGINE_PSOC] = {
-		.instances_dist = 0x1000,
-		.min_short_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
-		.max_short_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
-		.min_full_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_0,
-		.max_full_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_0,
-		.hit_wr_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AW,
-		.hit_rd_breg = mmJT_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
 	},
 	[HL_SEC_ENGINE_SFT] = {
 		.instances_dist = 0,
@@ -335,6 +333,68 @@
 	regmap_writel(hit | BIT(rr_idx), rr_dest);
 }
 
+#define BITS_IN_REG			32
+#define GAUDI2_HBW_SHORT_MASK		0xF
+#define GAUDI2_MAX_HBW_SHORT_RR		6
+#define GAUDI2_HBW_WRITE_ONLY_RR_IDX	8
+
+static void hl_secure_update_rr_reg_hbw(int rr_idx, u32 inst_offset,
+		struct hbw_rr_reg *rr_reg)
+{
+	u32 rr_base, rr_dest;
+	u32 data_max_hi = ((u32)(rr_reg->data_max >> BITS_IN_REG));
+	u32 data_min_hi = ((u32)(rr_reg->data_min >> BITS_IN_REG));
+	bool write_only = rr_idx >= GAUDI2_HBW_WRITE_ONLY_RR_IDX;
+
+	if (rr_idx < GAUDI2_MAX_HBW_SHORT_RR) {
+		data_max_hi &= GAUDI2_HBW_SHORT_MASK;
+		data_min_hi &= GAUDI2_HBW_SHORT_MASK;
+	}
+
+	rr_base = rr_reg->rr_idx_calc * sizeof(u32) + inst_offset;
+	rr_dest = rr_base + rr_reg->reg_min.lo;
+	regmap_writel((u32)rr_reg->data_min, rr_dest);
+	rr_dest = rr_base + rr_reg->reg_min.hi;
+	regmap_writel(data_min_hi, rr_dest);
+	rr_dest = rr_base + rr_reg->reg_max.lo;
+	regmap_writel((u32)rr_reg->data_max, rr_dest);
+	rr_dest = rr_base + rr_reg->reg_max.hi;
+	regmap_writel(data_max_hi, rr_dest);
+
+	hl_secure_update_hit_reg(rr_idx, inst_offset, &rr_reg->hit, write_only);
+}
+
+static void hl_secure_hbw_rr_gen(int rr_idx, u64 min, u64 max,
+		int dist, int inst_num, enum hl_sec_engine_type type)
+{
+	int i;
+	struct hbw_rr_reg rr_reg;
+	struct hl_sec_rrs_cfg_hbw *cfg = &hl_sec_hbw_rrs_tbl[type];
+
+	rr_reg.data_min = min >> SHIFT_4K_POS;
+	rr_reg.data_max = max >> SHIFT_4K_POS;
+
+	if (rr_idx < GAUDI2_MAX_HBW_SHORT_RR) {
+		rr_reg.rr_idx_calc = rr_idx;
+		rr_reg.reg_min = cfg->min_short;
+		rr_reg.reg_max = cfg->max_short;
+	} else {
+		rr_reg.rr_idx_calc = rr_idx - GAUDI2_MAX_HBW_SHORT_RR;
+		rr_reg.reg_min = cfg->min_full;
+		rr_reg.reg_max = cfg->max_full;
+	}
+
+	rr_reg.reg_min.lo += dist;
+	rr_reg.reg_min.hi += dist;
+	rr_reg.reg_max.lo += dist;
+	rr_reg.reg_max.hi += dist;
+	rr_reg.hit.hit_wr = cfg->hit_wr_breg + dist;
+	rr_reg.hit.hit_rd = cfg->hit_rd_breg + dist;
+
+	for (i = 0; i < inst_num; i++)
+		hl_secure_update_rr_reg_hbw(rr_idx, cfg->instances_dist * i, &rr_reg);
+}
+
 static void hl_secure_update_rr_reg(int rr_idx, u32 inst_offset, struct lbw_rr_reg *rr_reg)
 {
 	u32 rr_base, rr_dest;
@@ -403,7 +463,8 @@
 	}
 
 	if (rr_sel.psoc)
-		hl_secure_lbw_rr_gen(rr_idx, min, max, 0, PSOC_INST_NUM, HL_SEC_ENGINE_PSOC);
+		hl_secure_hbw_rr_gen(HL_SEC_HBW_PSOC_LBW, min, max, 0, PSOC_INST_NUM,
+					HL_SEC_ENGINE_PSOC);
 }
 
 #define EN_ARC_SEC_PRIV				0x3
@@ -529,10 +590,6 @@
 	memset(&rr_sel, 0, sizeof(struct hl_rr_eng_select));
 
 	/*
-	 * TODO:Add psoc rrs once determined for lbw.
-	 */
-
-	/*
 	 * In order to open psoc global space in preboot,
 	 * a new RR LBW_PSOC_GLBL_CONF is used to lock
 	 * global conf end to cfg space end and then the
@@ -549,6 +606,7 @@
 	 * reconfiguring in preboot.
 	 */
 	rr_sel.noc = true;
+	rr_sel.psoc = true;
 
 	/* CFG SPACE: 0x1000007FF8000000 - 0x1000007FFFFFFFFF */
 	min = CFG_BASE;
@@ -577,68 +635,6 @@
 	min = CFG_BASE;
 	max = mmPSOC_GLOBAL_CONF_BASE;
 	hl_secure_lbw_rr(HL_SEC_LBW_DBG_CFG_SPACE, min, max, rr_sel);
-}
-
-#define BITS_IN_REG			32
-#define GAUDI2_HBW_SHORT_MASK		0xF
-#define GAUDI2_MAX_HBW_SHORT_RR		6
-#define GAUDI2_HBW_WRITE_ONLY_RR_IDX	8
-
-static void hl_secure_update_rr_reg_hbw(int rr_idx, u32 inst_offset,
-		struct hbw_rr_reg *rr_reg)
-{
-	u32 rr_base, rr_dest;
-	u32 data_max_hi = ((u32)(rr_reg->data_max >> BITS_IN_REG));
-	u32 data_min_hi = ((u32)(rr_reg->data_min >> BITS_IN_REG));
-	bool write_only = rr_idx >= GAUDI2_HBW_WRITE_ONLY_RR_IDX;
-
-	if (rr_idx < GAUDI2_MAX_HBW_SHORT_RR) {
-		data_max_hi &= GAUDI2_HBW_SHORT_MASK;
-		data_min_hi &= GAUDI2_HBW_SHORT_MASK;
-	}
-
-	rr_base = rr_reg->rr_idx_calc * sizeof(u32) + inst_offset;
-	rr_dest = rr_base + rr_reg->reg_min.lo;
-	regmap_writel((u32)rr_reg->data_min, rr_dest);
-	rr_dest = rr_base + rr_reg->reg_min.hi;
-	regmap_writel(data_min_hi, rr_dest);
-	rr_dest = rr_base + rr_reg->reg_max.lo;
-	regmap_writel((u32)rr_reg->data_max, rr_dest);
-	rr_dest = rr_base + rr_reg->reg_max.hi;
-	regmap_writel(data_max_hi, rr_dest);
-
-	hl_secure_update_hit_reg(rr_idx, inst_offset, &rr_reg->hit, write_only);
-}
-
-static void hl_secure_hbw_rr_gen(int rr_idx, u64 min, u64 max,
-		int dist, int inst_num, enum hl_sec_engine_type type)
-{
-	int i;
-	struct hbw_rr_reg rr_reg;
-	struct hl_sec_rrs_cfg_hbw *cfg = &hl_sec_hbw_rrs_tbl[type];
-
-	rr_reg.data_min = min >> SHIFT_4K_POS;
-	rr_reg.data_max = max >> SHIFT_4K_POS;
-
-	if (rr_idx < GAUDI2_MAX_HBW_SHORT_RR) {
-		rr_reg.rr_idx_calc = rr_idx;
-		rr_reg.reg_min = cfg->min_short;
-		rr_reg.reg_max = cfg->max_short;
-	} else {
-		rr_reg.rr_idx_calc = rr_idx - GAUDI2_MAX_HBW_SHORT_RR;
-		rr_reg.reg_min = cfg->min_full;
-		rr_reg.reg_max = cfg->max_full;
-	}
-
-	rr_reg.reg_min.lo += dist;
-	rr_reg.reg_min.hi += dist;
-	rr_reg.reg_max.lo += dist;
-	rr_reg.reg_max.hi += dist;
-	rr_reg.hit.hit_wr = cfg->hit_wr_breg + dist;
-	rr_reg.hit.hit_rd = cfg->hit_rd_breg + dist;
-
-	for (i = 0; i < inst_num; i++)
-		hl_secure_update_rr_reg_hbw(rr_idx, cfg->instances_dist * i, &rr_reg);
 }
 
 static void hl_secure_hbw_rr(int rr_idx, u64 min, u64 max, struct hl_rr_eng_select rr_sel)
